thorlcr/master/thgraphmanager.cpp

```diff
@@ -60,10 +61,163 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+                sock->cancel_accept();
+            threaded.join();
+        }
+        virtual unsigned getPort() { return port; }
+
+        virtual void processDebugCommand(CSafeSocket &ssock, StringBuffer &rawText)
+        {
+            Owned<IPropertyTree> queryXml;
+            try
+            {
+                queryXml.setown(createPTreeFromXMLString(rawText.str(), ipt_caseInsensitive, (PTreeReaderOptions)(ptr_ignoreWhiteSpace|ptr_ignoreNameSpaces)));
+            }
+            catch (IException *E)
+            {
+                StringBuffer s;
+                DBGLOG("ERROR: Invalid XML received from %s:%s", E->errorMessage(s).str(), rawText.str());
+                throw;
+            }
+
+            Linked<CJobMaster> job = mgr.getCurrentJob();
+            const char *graphId = job->queryGraphName();
+            if (!job || !graphId)
+                throw MakeStringException(5300, "Command not available when no graph active");
+
+            const char *command = queryXml->queryName();
+            if (!command) throw MakeStringException(5300, "Invalid debug command");
+
+            FlushingStringBuffer response(&ssock, false, MarkupFmt_XML, false, false, queryDummyContextLogger());
+            response.startDataset("Debug", NULL, (unsigned) -1);
+
+            if (strncmp(command,"print", 5) == 0)
+            {
+                const char *edgeId = queryXml->queryProp("@edgeId");
+                if (!edgeId) throw MakeStringException(5300, "Invalid edgeId command");
+
+                ICommunicator &comm = job->queryNodeComm();
+                CMessageBuffer mbuf;
+                mbuf.append(DebugRequest);
+                mbuf.append(job->queryKey());
+                mptag_t replyTag = createReplyTag();
+                serializeMPtag(mbuf, replyTag);
+                mbuf.append(rawText);
+                if (!comm.send(mbuf, RANK_ALL_OTHER, masterSlaveMpTag, MP_ASYNC_SEND))
+                {
+                    DBGLOG("Failed to send debug info to slave");
+                    throwUnexpected();
+                }
+                unsigned nodes = job->queryNodes();
+                response.appendf("<print graphId='%s' edgeId='%s'>", graphId, edgeId);
+                while (nodes)
+                {
+                    rank_t sender;
+                    mbuf.clear();
+                    comm.recv(mbuf, RANK_ALL, replyTag, &sender, 10000);
+                    while (mbuf.remaining())
+                    {
+                        StringAttr row;
+                        mbuf.read(row);
+                        response.append(row);
+                    }
+                    nodes--;
+                }
+                response.append("</print>");
+            }
+            else if (strncmp(command,"quit", 4) == 0)
+            {
+                LOG(MCwarning, thorJob, "ABORT detected from user");
+                Owned <IException> e = MakeThorException(TE_WorkUnitAborting, "User signalled abort");
```

Comment:
Note, should I also modify the original code where there is a space there also?  (thgraphmaster.cpp line 1637)